### DISCUSSION TITLE
Paps ideas

### DISCUSSION THREAD

**Discussion by dov** [dov/paps#45] [General]
Hi all,

Please raise ideas and general feedback here in the discussion forum.

/Dov

**Comment by gnusupport** [comment]
`paps` is very useful utility for reason that it can print Unicode chars and Emoji without problems. For many types of documents there is no need to complicate with various markups when simple plain text is sufficient. 

There are other similar utilities, but they do not and did not, for decades, support Unicode characters. That is major problem that I find.

Generating PDF files is difficult without Unicode in ever evolving world of 2023.

Maybe `paps` could learn from other utilities, such as `a2ps` which does not have Unicode support and `enscript`. Bug has been sent. Though I have little hopes that it will ever be implemented.

See the options for ` a2ps`:
https://gnu.support/files/programs/a2ps/a2ps.pdf
It supports various encodings, but not full Unicode.

See the options for `enscript`:
https://gnu.support/files/programs/enscript/enscript.pdf
I don't think it support encodings.

Those cards have been prepared with `card` utility that comes with `a2ps` package.

Both programs have excellent features. Such as having the frame on the page, or nicer, more visible title.

**Comment by gnusupport** [comment]
Another useful idea is Pango markup itself, as `paps` already supports such markup, and that would mean that meta-level markup could be prepared to allow various outputs.

For example, in Emacs, there is Org mode, and Org mode supports now hundred(s) various outputs. But not Pango markup, it would be good idea to create Pango markup output for Org mode and then use Paps to generate .ps and .pdf files.

Common Lisp already supports Pango markup:

Pango Markup:
https://shinmera.github.io/pango-markup/

I am not sure if the following is the reference to Pango markup:

Pango – 1.0: Text Attributes and Markup:
https://docs.gtk.org/Pango/pango_markup.html

Emacs has the Enriched text mode, which could then be used to translate to Pango markup and get fancy PDF files.

**Reply by gnusupport** [reply]
It was my English, I wanted to say

> Another useful idea is Pango markup itself, as `paps` already supports such markup, and that would mean that meta-level markup could be prepared to allow various outputs.

Yes, I know pango markup is supported.

Emacs has invisible properties, so letters can be for example justified, or be in various colors. That can be translated to pango markup because `paps` supports it. And so translation could take place into HTML or Pango or other formats as well.

It would require meta-level markup, something like Markdown, but more complex to be able to include `span` attributes.

**Comment by gnusupport** [comment]
A feature I would like to have is possibility to include picture as full background picture of the page, including different background pictures for different pages. Sounds complex, I know. But that alone would enable creation of brochures and presentation materials.

Imagine a letterhead design as background. For visual understanding, see here: https://duckduckgo.com/?t=h_&q=letterheard&ia=web

If `paps` would have option to include background over full page, it would make the tool significant.

Maybe options could be as following:

`--bgimage-for-pages=page1=1.jpg,page2=2.jpg` 

and such option would overwrite any other option, such as following:

`--front-bgimage=1.jpg` 
`--back-bgimage=2.jpg`
`--page-bgimage=page.jpg`

How would logic go?

If `--page-bgimage=page.jpg` is set alone, the image `page.jpg` would appear everywhere.

If `--back-bgimage=2.jpg` is set alone, that image would appear only on the last page, overwriting `--page-bgimage=page.jpg` setting.

If `--front-bgimage=1.jpg` is set alone, that image would appear only on the first page, overwriting `--page-bgimage=page.jpg` setting.

And finally `--bgimage-for-pages=page1=1.jpg,page2=2.jpg` would overwrite background images, if specified as `page1=1.jpg` or `front=1.jpg` and `back=back.jpg` and any other page with setting such as `page%d=%f.jpg`. Using `back=background.jpg` is more conclusive for program rather than page number.

**Comment by gnusupport** [comment]
I have already created few documents by RCD Paps for Emacs using underlying `paps` as translation papers which I will bring to https://www.mak.ac.ug/ Makerere University for certification. This way you know that `paps` was used in significant activity influencing many people's lives.

**Comment by dov** [comment]
@gnusupport Thanks for the ideas! Note that markup is already supported. Just give the `--markup` option.

**Reply by gnusupport** [reply]
I seriously consider converting Emacs enriched text to Pango markup for output. That would mean visual editing, colors, background colors, bold, italic, etc. and then getting output on paper or in PDF by using `paps`. I hope you get the idea. 

But I am thinking that including images would be great. But how? I have no idea. With images and markup, it would be good solution for document  creation. Right now is good for text only.

**Comment by dov** [comment]
I  think that is outside the scope of paps. If you are using org-mode, which is one of the typical use cases for creating documents from emacs, you can already export to pdf through LaTeX or via html.

Your idea about static images below the text does make sense though.

**Reply by gnusupport** [reply]
`paps` is for PDF, printing, or final result. Emacs enriched mode is very different to Org mode. It allows visual, similar to WYSIWYG principle text writing.

**Reply by mintty** [reply]
I don't mean to interpret the scope of paps, just compare to text terminals which have a much smaller scope concerning text formatting features, yet there are a number of graphics features implemented in a couple of them: Regis, Sixel, Tek, iTerm2 image protocol.

**Reply by dov** [reply]
If, and I dont say I will do it, paps were to be extended to include images we would need to interpret some kind of markup language. pango markup is obviously not enough. I guess we could extend it with an <image> attribute to create a paps flavored pango markup language.  Or do you have an idea of another language? Does emacs allow serializing its buffers including images? Perhaps it would be better to merge paps into emacs?

**Reply by gnusupport** [reply]
> `paps` is for PDF, printing, or final result. Emacs enriched mode is very different to Org mode. It allows visual, similar to WYSIWYG principle text writing.

It is not hard to make Org export to Pango markup, that is easiest and will be beneficial.

But Pango markup is so much more visually expressive, not though typesetting-like expressive. However, that simplicity is really good enough for many applications like invoices, listings, balance sheets, etc.

**Reply by gnusupport** [reply]
I have asked Pandoc author to make Pango output/input, but not answer yet. It is not popular, but is obviously visually very nice. 

Regarding Emacs and any other editor, I am thinking that Pango and paps together, should be used in such way that what is visually on screen, that "about the same" come out on paper.

It is also possible to think of making Paps mode for Emacs, where one could use those attributes rather related to Pango markup, but in visual manner WYSIWYG.

It is possible to make translators, for example for txt2tags to translate some paps mode to Pango markup and others, that works. It is possible to make Markdown exporter. Even easier would be with Djot:

jgm/djot: A light markup language:
https://github.com/jgm/djot#rationale

As that one can use Pandoc for conversions to/from many formats. It is similar to Markdown, more expressive and not dependent to HTML.

I am rather thinking of having visual WYSIWYG editing that comes out on PDF or paper by using Pango and paps.

For images, if one uses Emacs package `telega` one could see that images could be placed left or right from text, but such images are split to the height of the text line. I do not believe that this technique should be employed in Paps.

What I think is that images could be included in single line, left, center, or right aligned. That would be enough. It requires few parameters for width, height, maybe caption. 

Images can be used for example, for logo, letter head, header as image, footer as image, or embedding images for purposes of presentation.

**Comment by EPRINC-MP** [comment]
I run the command:
paps --columns=3 --landscape --paper=letter --font="Palatino-Roman 6" /path/to/some/textfile | lpr -P SomePrinterName

It produces good results; the command was suggested to me using "palatino" for a fontname.

One irony is that when I search I do not find a palatino file on my computers. 

Enscript has a file of usable fontnames; the list is viewable by issuing the following command:
cat /usr/share/enscript/afm/font.map

Is there a way of listing which fonts/fontnames are usable by paps?

Much thanks, 

Max

**Reply by dov** [reply]
paps is using pango, which is using font-config for finding fonts. To get a list of font-config fonts do:
```
fc-list
```
Please refer to the fc-list man page for more info.

I should probably mention this in the man page.
